jupyterhub_singleuser_profiles/service.py

```diff
@@ -37,91 +41,110 @@ def __init__(self, server_url, token, namespace=None, verify_ssl=True):
+        }
+      config_map_wrapper.create(body=body, namespace=self.namespace)
+      result = config_map_wrapper.get(namespace=self.namespace, name=_REFERENCE_CM_NAME %(escape(user)))
+      _LOGGER.debug("Created reference config map")
+      return result
+    except ApiException as e:
+      if e.status == 409:
+        try:
+          result = config_map_wrapper.get(namespace=self.namespace, name=_REFERENCE_CM_NAME %(escape(user)))
+          _LOGGER.debug("Got reference config map")
+          return result
+        except ApiException as e2:
+          _LOGGER.error("Error getting reference ConfigMap in %s for %s: %s\n" % (self.namespace, escape(user), e2))
+          raise
+      else:
+        _LOGGER.error("Error creating reference ConfigMap in %s for %s: %s\n" % (self.namespace, escape(user), e))
+        raise
 
+  def get_template(self, name, path):
+    cm = self.os_client.resources.get(api_version='v1', kind='ConfigMap')
     try:
-      response = tmplt.get(
+      response = cm.get(
           namespace=self.namespace,
           name=name
       )
-      template = response.to_dict()
+      cm = response.to_dict()
     except Exception as e:
       _LOGGER.error("Error: %s %s" % (name, e))
       template = None
+    template = yaml.load(cm['data'].get(path))
 
     return template
 
-  def process_template(self, user, template, **parameters):
-
-    self._set_template_parameters(
-      template,
-      USERNAME=escape(user),
-      **parameters
-    )
-    template["apiVersion"] = "template.openshift.io/v1"
-
-    endpoint = "{}/apis/template.openshift.io/v1/namespaces/{}/processedtemplates".format(
-        self.server_url,
-        self.namespace
-    )
-    response = requests.post(
-        endpoint,
-        json=template,
-        headers={
-            'Authorization': 'Bearer {}'.format(self.token),
-            'Content-Type': 'application/json'
-        },
-        verify=False #FIXME?
-    )
-    _LOGGER.debug("OpenShift master response template (%d): %r", response.status_code, response.text)
-
-    try:
-        response.raise_for_status()
-    except Exception:
-        _LOGGER.error("Failed to process template: %s", response.text)
-        raise
+  def process_template(self, user, service_name, template, configuration, labels=None):
 
-    processed = response.json()
-    if len(processed["objects"]) != 1:
-      raise Exception("Template must contain a single ConfigMap or CustomResource")
+    tmp = jinja2.Template(json.dumps(template))
 
-    configMap = processed["objects"][0]
-    if not configMap["metadata"].get("labels"):
-      configMap["metadata"]["labels"] = {}
+    result = tmp.render(configuration)
 
-    configMap["metadata"]["labels"][_SERVICE_LABEL] = escape(user)
+    result = json.loads(result)
+    if not result.get('metadata'):
+      result['metadata'] = {}
+    if labels:
+      result['metadata']['labels'] = labels
+    
+    result['metadata']['name'] = service_name + '-' + user
```

Comment:
I am thinking if we should let user parameterize the name. I.e suggest in examples that the name should be something like `service-name-{{ user }}` and then add `user` to `render` call, so that we keep this flexible.

We might add a check and enforce the name if the user does not use the parameterization

 `if user not in result['metadata']['name']: set the name as it is done now`

what do you think?

Reply:
I will put the username as part of the configuration dict and check after render if the username was specified. If it was not it will rewrite that part of the already rendered cm.